Update Go to 1.19

tklauser opened the pull request:
Release blog: https://go.dev/blog/go1.19
Release notes: https://go.dev/doc/go1.19

tklauser committed:
all: gofmt with Go 1.19

Go 1.19 gofmt changes how doc comments are formatted [0], so reformat
all files so that they are compatible with the new format.

Adjust some comments where reformatting would break existing godoc
comments.

[0] https://go.dev/doc/go1.19#go-doc

Signed-off-by: Tobias Klauser <tobias@cilium.io>

tklauser committed:
ci: update golangci-lint to 1.48.0

Signed-off-by: Tobias Klauser <tobias@cilium.io>

sayboras commented on plugins/cilium-cni/main.go:
```diff
@@ -3,7 +3,7 @@
 // Ensure build fails on versions of Go that are not supported by Cilium.
 // This build tag should be kept in sync with the version specified in go.mod.
-//go:build go1.18
+//go:build go1.19
```
any reason to rename this file to main.go ?

tklauser replied:
Sorry, I forgot to mention this in the commit message. It's so this bit of `update-go-version` works for `plugins/cilium-cni` as well:

https://github.com/cilium/cilium/blob/09a68d6a5f5db01f8b73debce181f79d798fb8f2/Makefile#L615-L621

All other binaries' `main` funcs are already in `main.go` files, but this was probably missed.

sayboras replied:
oh right, all good :100:

sayboras reviewed (approved):
:rocket: 

Just a question for my understanding, we will update dev VM and CI jenkins soon as well right ?

tklauser committed:
Update Go to 1.19

Release blog: https://go.dev/blog/go1.19
Release notes: https://go.dev/doc/go1.19

Signed-off-by: Tobias Klauser <tobias@cilium.io>

qmonnet commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -192,11 +192,12 @@ func uniqueCESliceName(desiredCESs *CESToCEPMapping) string {
-// 2) During operator warm boot [after crash or software upgrade], slicing manager
-//    creates a CES, by passing unique name.
+//
+//		called on 2 different scenarios.
```
This one might need a small adjustment, it should be aligned like the sentence above and not like the items below?

qmonnet commented on operator/pkg/ingress/ingress.go:
```diff
@@ -52,11 +52,11 @@ type ingressServiceUpdatedEvent struct {
+//   - CiliumEnvoyConfig
 //   - Manage synced TLS secrets in given namespace
-//		- TLS secrets
+//   - TLS secrets
```
The different indent level was probably on purpose, and seems lost here?

tklauser replied:
Unfortunately, it looks like `gofmt` doesn't support nested lists :(

> List items only contain paragraphs, not code blocks or nested lists.

https://go.dev/doc/comment#lists

If rewritten the comment to use a mix of numbered and unnumbered lists as suggested in https://go.dev/doc/comment#mistakes.

qmonnet commented on pkg/allocator/allocator.go:
```diff
@@ -49,24 +49,25 @@ const (
-//       etcd: Create is made conditional on existence of master key
-//       consul: locking
+//
+//	key can be removed in the meantime.
```
The new line breaks the sentence

qmonnet commented on pkg/azure/types/types.go:
```diff
@@ -44,7 +44,7 @@ type AzureSpec struct {
 // AzureStatus is the status of Azure addressing of the node
 //
-// This struct is embedded into v2.CiliumNode
+// # This struct is embedded into v2.CiliumNode
```
Not sure where the `#` comes from :thinking:

tklauser replied:
`gofmt` probably interprets this as the start of a list due to the empty line above :thinking:

qmonnet commented on pkg/crypto/certloader/certloader_test.go:
```diff
@@ -366,19 +366,19 @@ func k8sDataDirName() string {
+//   - key: tls.key
+//     path: server.key
+//     name: hubble-server-certs
+//     optional: true
```
We lose the YAML syntax here!

tklauser replied:
> A code block is a span of indented or blank lines not starting with a bullet list marker or numbered list marker. It is rendered as preformatted text (a &lt;pre&gt; block in HTML).

-- https://go.dev/doc/comment#code

:sob:

qmonnet commented on pkg/datapath/linux/linux_defaults/mark.go:
```diff
@@ -7,17 +7,20 @@ package linux_defaults
 // following way:
 //
-//  1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
+//	1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2
```
The additional spaces throw the indices off with the rest of the diagrams, here and below :(

qmonnet commented on pkg/endpoint/events.go:
```diff
@@ -109,7 +109,7 @@ func (e *Endpoint) PolicyRevisionBumpEvent(rev uint64) {
 }
 
-/// EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
+// / EndpointNoTrackEvent contains all fields necessary to update the NOTRACK rules.
```
Let's get rid of this `/`?

qmonnet commented on pkg/fqdn/cache.go:
```diff
@@ -803,7 +805,8 @@ func (zombies *DNSZombieMappings) isConnectionAlive(zombie *DNSZombieMapping) bo
 // getAliveNames returns all the names that are alive
-//   a name is alive if at least one of the IPs that resolve to it is alive
+//
+//	a name is alive if at least one of the IPs that resolve to it is alive
```
Should be the same paragraph as above?

qmonnet commented on pkg/fqdn/doc.go:
```diff
@@ -34,8 +34,10 @@
+//
+//	|          +----------------+   Data      ^               ^
+//	v                                         |               |
+//
```
Ouch

qmonnet commented on pkg/fqdn/doc.go:
```diff
@@ -53,15 +55,15 @@
+//     rules:
+//     dns:
+//   - matchPattern: "*"
+//   - matchName: "cilium.io"
```
YAML indent lost

qmonnet commented on pkg/fqdn/doc.go:
```diff
@@ -92,9 +94,9 @@
-//      - matchPattern: "bucket.*.my-remote-service.com"
+//   - toFQDNs:
+//   - matchName: "my-remote-service.com"
+//   - matchPattern: "bucket.*.my-remote-service.com"
```
YAML indent lost

qmonnet commented on pkg/fqdn/doc.go:
```diff
@@ -190,36 +192,47 @@
+//
+//	|
+//	v
+//
```
All diagrams above get broken

qmonnet commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -883,15 +883,15 @@ spec:
+                    Requires field in any     rule will apply to all other rules as
+                    well. \n   - ToEndpoints, ToCIDR, ToCIDRSet, ToEntities, ToServices
+                    and ToGroups are     mutually exclusive. Only one of these members
+                    may be present within an     individual rule."
```
Some weird slashes and space in this paragraph, but I suppose this is auto-generated?

tklauser replied:
Yes, this is auto-generated and contains the respective godoc comment (which was changed in the corresponding Go file this was generated from).

qmonnet commented on pkg/k8s/slim/k8s/apis/labels/selector.go:
```diff
@@ -138,7 +138,8 @@ type Requirement struct {
 // (6) The key is invalid due to its length, or sequence
-//     of characters. See validateLabelKey for more details.
+//
+//	of characters. See validateLabelKey for more details.
```
This should be part of the item above. Probably `(n)` numbers that need to be turned into `n.`.

qmonnet commented on pkg/k8s/slim/k8s/apis/labels/selector.go:
```diff
@@ -197,13 +198,20 @@ func (r *Requirement) hasValue(value string) bool {
 // (5) The operator is GreaterThanOperator or LessThanOperator, and Labels has
-//     the Requirement's key and the corresponding value satisfies mathematical inequality.
+//
+//	the Requirement's key and the corresponding value satisfies mathematical inequality.
```
Same as above.

qmonnet commented on pkg/k8s/slim/k8s/apis/labels/selector.go:
```diff
@@ -828,32 +836,33 @@ func (p *Parser) parseExactValue() (sets.String, error) {
+//	(3) The empty string is a valid VALUE
+//	(4) A requirement with just a KEY - as in "y" above - denotes that
+//	    the KEY exists and can be any VALUE.
+//	(5) A requirement with just !KEY requires that the KEY not exist.
```
`n.` as well, also it seems the existing indent has preserved this paragraph from being changed.

qmonnet commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -28,26 +28,29 @@ var (
+//     Master keys provide the mapping from ID to key. As long as a master key
+//     for an ID exists, the ID is still in use. However, if a master key is no
+//     longer backed by at least one slave key, the garbage collector will
+//     eventually release the master key and return it back to the pool.
```
The last paragraph should not be indented like the items above?

qmonnet commented on pkg/policy/l4.go:
```diff
@@ -316,10 +316,11 @@ func (l4 *L4Filter) GetPort() uint16 {
+//   - Entry with ProxyPort = 0: No proxy redirection is needed for this key
+//   - Entry with any other port #: Proxy redirection is required for this key,
+//     caller must replace the ProxyPort with the actual
+//     listening port number.
```
We could remove the `-` at the beginning to have a description list and keep the indent clearer

qmonnet commented on pkg/policy/l4_filter_test.go:
```diff
@@ -1555,8 +1555,9 @@ func (ds *PolicyTestSuite) TestMergingWithDifferentEndpointSelectedAllowAllL7(c
-//          allow all; other L3 should restrict at L7 in a separate filter.
+//
+//	the host should always be allowed. From Host should go to proxy
+//	allow all; other L3 should restrict at L7 in a separate filter.
```
Broken paragraph

qmonnet reviewed (changes requested):
Thanks a lot! I'm afraid the `gofmt` changes break the comments at a number of locations and will require some prior adjustments, I've commented below on everything I found.

qmonnet reviewed (approved):
> Thanks for the detailed godoc review @qmo. I've tried to resolve all locations where comments were broken by `gofmt`.

Thanks!

> FWIW, for YAML code block I've had to resort to prefixing them with `---` because otherwise they would always be treated as lists :( see https://github.com/cilium/cilium/pull/20816#discussion_r941122932

The `---` looks absolutely fine to me.

Nice work, thanks for addressing it all! I can't really exclude missing things because of the number of formatting changes in the PR, but still I did a second quick pass and didn't notice any broken comment this time. Looks all good!

gandro commented on api/v1/client/daemon/daemon_client.go:
```diff
@@ -48,7 +48,7 @@ type ClientService interface {
 }
 
 /*
-  GetClusterNodes gets nodes information stored in the cilium agent
```
Almost all `.go` files in `api/v1` are auto-generated. Are these formatting changes due to the builder base image change or did you explicitly gofmt those files? If the latter, then they'll likely get undone if you run `make -C api/v1` again.

tklauser replied:
These got `gofmt`ed like that when running `contrib/scripts/check-api-code-gen.sh` (which in turn runs `make -C api/v1`) due to the builder image being update to Go 1.19, it seems. This is also what's run as part of CI in the `generate-api` step of the `lint-go` workflow.

aanm reviewed (changes requested):
Sebastian brought a good point with regarding the auto-generated files. Perhaps we should run `gofmt` after we generate the files similar that is done with goimports in here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L321-L324
here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L341 and here https://github.com/cilium/cilium/blob/9e6f2bb9ee6c6811583c42b239aaccb15eb998de/Makefile#L358
